overlord/state/warning.go

```diff
@@ -0,0 +1,282 @@
+	return w.message
+}
+
+func (w *Warning) MarshalJSON() ([]byte, error) {
+	jw := jsonWarning{
+		Message:     w.message,
+		FirstAdded:  w.firstAdded,
+		LastAdded:   w.lastAdded,
+		ExpireAfter: w.expireAfter.String(),
+		RepeatAfter: w.repeatAfter.String(),
+	}
+	if !w.lastShown.IsZero() {
+		jw.LastShown = &w.lastShown
+	}
+
+	return json.Marshal(jw)
+}
+
+func (w *Warning) UnmarshalJSON(data []byte) error {
+	var jw jsonWarning
+	err := json.Unmarshal(data, &jw)
+	if err != nil {
+		return err
+	}
+	w.message = jw.Message
+	w.firstAdded = jw.FirstAdded
+	w.lastAdded = jw.LastAdded
+	if jw.LastShown != nil {
+		w.lastShown = *jw.LastShown
+	}
+	if jw.ExpireAfter != "" {
+		w.expireAfter, err = time.ParseDuration(jw.ExpireAfter)
+		if err != nil {
+			return err
+		}
+	}
+	if jw.RepeatAfter != "" {
+		w.repeatAfter, err = time.ParseDuration(jw.RepeatAfter)
+		if err != nil {
+			return err
+		}
+	}
+
+	return w.validate()
+}
+
+func (w *Warning) validate() (e error) {
+	if w.message == "" {
+		return errNoMessage
+	}
+	if strings.TrimSpace(w.message) != w.message {
+		return errBadMessage
+	}
+	if w.firstAdded.IsZero() {
+		return errNoFirstAdded
+	}
+	if w.expireAfter == 0 {
+		return errNoExpireAfter
+	}
+	if w.repeatAfter == 0 {
+		return errNoRepeatAfter
+	}
+	return nil
+}
+
+func (w *Warning) ExpiredBefore(now time.Time) bool {
+	return w.lastAdded.Add(w.expireAfter).Before(now)
+}
+
+func (w *Warning) ShowAfter(t time.Time) bool {
+	return (w.lastShown.IsZero() || w.lastShown.Add(w.repeatAfter).Before(t)) && !w.firstAdded.After(t)
+}
+
+// flattenWarning loops over the warnings map, and returns all
+// non-expired warnings therein as a flat list, for serialising.
+// Call with the lock held.
+func (s *State) flattenWarnings() []*Warning {
+	now := time.Now()
+	flat := make([]*Warning, 0, len(s.warnings))
+	for _, w := range s.warnings {
+		if w.ExpiredBefore(now) {
+			continue
+		}
+		flat = append(flat, w)
+	}
+	return flat
+}
+
+// unflattenWarnings takes a flat list of warnings and replaces the
+// warning map with them, ignoring expired warnings in the process.
+// Call with the lock held.
+func (s *State) unflattenWarnings(flat []*Warning) {
+	now := time.Now()
+	s.warnings = make(map[string]*Warning, len(flat))
+	for _, w := range flat {
+		if w.ExpiredBefore(now) {
+			continue
+		}
+		s.warnings[w.message] = w
+	}
+}
+
+// Warnf records a warning: if it's the first Warning with this
+// message it'll be added (with its firstAdded and lastAdded set to the
+// current time), otherwise the existing one will have its lastAdded
+// updated.
+func (s *State) Warnf(template string, args ...interface{}) {
+	var message string
+	if len(args) > 0 {
+		message = fmt.Sprintf(template, args...)
+	} else {
+		message = template
+	}
+	s.addWarningFull(Warning{
+		message:     message,
+		expireAfter: DefaultExpireAfter,
+		repeatAfter: DefaultRepeatAfter,
+	}, time.Now().UTC())
+}
+
+func (s *State) addWarningFull(w Warning, t time.Time) {
```

Comment:
should this still have the Full in the name given we don't have AddWarning itself anymore ?